### PR TITLE
remove unnecessary internal functions and remove loans.rs, util.rs

### PR DESCRIPTION
**Pull request by yz89**
related #30 #107

### CODE REVIEW

**Review by GopherJ — APPROVED** [review]

File: pallets/loans/src/lib.rs [hunk 1]
```diff
@@ -699,6 +737,439 @@ impl<T: Config> Pallet<T> {
...
+        Ok(())
+    }
+
+    fn collateral_asset_internal(
+        who: T::AccountId,
+        currency_id: CurrencyId,
+        enable: bool,
+    ) -> result::Result<(), Error<T>> {
+        let mut collateral_assets = AccountCollateralAssets::<T>::get(&who);
+        if enable {
```
Comment (kaichaosun) [inline pallets/loans/src/lib.rs:921]: return error early to remove nested if else.

File: pallets/loans/src/lib.rs [hunk 2]
```diff
@@ -461,11 +461,31 @@ pub mod module {
...
+            )?;
+            TotalSupply::<T>::try_mutate(&currency_id, |total_balance| -> DispatchResult {
+                let new_balance = total_balance
+                    .checked_add(collateral_amount)
+                    .ok_or(Error::<T>::CollateralOverflow)?;
+                *total_balance = new_balance;
+                Ok(())
+            })?;
+            T::Currency::transfer(currency_id, &who, &Self::account_id(), mint_amount)?;
+
```
[the remark this answers is not in the document]
Comment (yz89) [reply pallets/loans/src/lib.rs:498]: I just added the essential events for all dispatchables.

File: pallets/loans/src/lib.rs [hunk 3]
```diff
@@ -516,10 +536,28 @@ pub mod module {
...
             let who = ensure_signed(origin)?;
-
             Self::ensure_currency(&currency_id)?;
 
-            Self::borrow_internal(&who, &currency_id, borrow_amount)?;
+            Self::borrow_guard(&who, &currency_id, borrow_amount)?;
+            let account_borrows = Self::borrow_balance_stored(&who, &currency_id)?;
+            let account_borrows_new = account_borrows
+                .checked_add(borrow_amount)
+                .ok_or(Error::<T>::CalcBorrowBalanceFailed)?;
```
[the remark this answers is not in the document]
Comment (yz89) [reply pallets/loans/src/lib.rs]: Solved

File: pallets/loans/src/lib.rs (hunk 1 above)
Comment (yz89) [reply pallets/loans/src/lib.rs:917]: I will refactor this after merge `AccountCollateralAssets` into `AccountCollateral`.

File: pallets/loans/src/lib.rs [hunk 4]
```diff
@@ -516,10 +536,28 @@ pub mod module {
...
+            let total_borrows = Self::total_borrows(&currency_id);
+            let total_borrows_new = total_borrows
+                .checked_add(borrow_amount)
+                .ok_or(Error::<T>::CalcBorrowBalanceFailed)?;
+            T::Currency::transfer(currency_id, &Self::account_id(), &who, borrow_amount)?;
```
[the remark this answers is not in the document]
Comment (yz89) [reply pallets/loans/src/lib.rs]: Solved

File: pallets/loans/src/lib.rs [hunk 5]
```diff
@@ -140,48 +122,54 @@ pub mod module {
...
         CollateralAssetAdded(T::AccountId, CurrencyId),
         /// Disable collateral for certain asset
         /// [sender, currency_id]
         CollateralAssetRemoved(T::AccountId, CurrencyId),
-
-        // TODO: add event for dispatchables `mint, redeem, borrow, repay, liquidate` (#32)
-        /// Liquidation occurred
-        /// [liquidator, borrower,liquidate_token,collateral_token,liquidate_token_repay_amount,collateral_token_amount]
-        LiquidationOccur(
+        /// Event emitted when asserts are deposited
```
Comment (kaichaosun) [inline pallets/loans/src/lib.rs:148]:
```suggestion
        /// Event emitted when assets are deposited
```

File: pallets/loans/src/lib.rs [hunk 6]
```diff
@@ -140,48 +122,54 @@ pub mod module {
...
+        /// [sender, currency_id, amount]
+        Deposit(T::AccountId, CurrencyId, Balance),
+        /// Event emitted when asserts are redeemed
```
Comment (kaichaosun) [inline pallets/loans/src/lib.rs:151]:
```suggestion
        /// Event emitted when assets are redeemed
```

**Review by kaichaosun — APPROVED** [review]
💯

**Pull request merged**